File: 2141-maximum-running-time-of-n-computers/2141-maximum-running-time-of-n-computers.py
class Solution:
    def maxRunTime(self, n: int, batteries: List[int]) -> int:
        batteries.sort()
        
        def check(mins):
            total = 0
            computers = 0
            for time in batteries:
                total += time
                if total >= mins:
                    computers += 1
                    total -= mins
            
            return computers >= n
        
        left = 0
        right = 10**18
        
        while left < right:
            mid = left + right >> 1
            
            if check(mid):
                left = mid + 1
            else:
                right = mid
                
        return left - 1
            

File: 2141-maximum-running-time-of-n-computers/2141-maximum-running-time-of-n-computers.java
class Solution {
    public long maxRunTime(int n, int[] batteries) {
        long sumPower = 0;
        for (int power : batteries)
            sumPower += power;
        long left = 1, right = sumPower / n;
        
         while (left < right){
            long target = right - (right - left) / 2;
            long extra = 0;
            
            for (int power : batteries)
                extra += Math.min(power, target);

            if (extra >= (long)(n * target))
                left = target;
            else
                right = target - 1;
        }
        return left;
    }
    
    boolean check(long mins, int[] batteries, int n) {
        long total = 0;
        long computers = 0;
        for (int time : batteries) {
            total += time;
            if (total >= mins) {
                computers++;
                total -= mins;
            }
        }
        return computers >= n;
    }
}